cmd/modelcmd/modelcommand.go

```diff
@@ -307,37 +314,84 @@ func (c *ModelCommandBase) NewAPIClient() (*api.Client, error) {
 	return root.Client(), nil
 }
 
+// ModelDetails returns details from the file store for the model indicated by
+// the currently set controller name and model identifier.
 func (c *ModelCommandBase) ModelDetails() (string, *jujuclient.ModelDetails, error) {
-	modelName, err := c.ModelName()
+	modelIdentifier, err := c.ModelIdentifier()
 	if err != nil {
 		return "", nil, errors.Trace(err)
 	}
 	controllerName, err := c.ControllerName()
 	if err != nil {
 		return "", nil, errors.Trace(err)
 	}
-	details, err := c.modelDetails(controllerName, modelName)
-	return modelName, details, err
+
+	name, details, err := c.modelDetails(controllerName, modelIdentifier)
+	return name, details, errors.Trace(err)
 }
 
-func (c *ModelCommandBase) modelDetails(controllerName, modelName string) (*jujuclient.ModelDetails, error) {
-	if modelName == "" {
-		return nil, errors.Trace(ErrNoModelSpecified)
+func (c *ModelCommandBase) modelDetails(controllerName, modelIdentifier string) (
+	string, *jujuclient.ModelDetails, error,
+) {
+	if modelIdentifier == "" {
+		return "", nil, errors.Trace(ErrNoModelSpecified)
 	}
-	details, err := c.store.ModelByName(controllerName, modelName)
+
+	name, details, err := c.modelFromStore(controllerName, modelIdentifier)
 	if err != nil {
 		if !errors.IsNotFound(err) {
-			return nil, errors.Trace(err)
+			return "", nil, errors.Trace(err)
 		}
-		logger.Debugf("model %q not found, refreshing", modelName)
-		// The model isn't known locally, so query the models
+		logger.Debugf("model %q not found, refreshing", modelIdentifier)
+		// The model is not known locally, so query the models
 		// available in the controller, and cache them locally.
 		if err := c.RefreshModels(c.store, controllerName); err != nil {
-			return nil, errors.Annotate(err, "refreshing models")
+			return "", nil, errors.Annotate(err, "refreshing models")
+		}
+		name, details, err = c.modelFromStore(controllerName, modelIdentifier)
+	}
+	return name, details, errors.Trace(err)
+}
+
+// modelFromStore attempts to retrieve details from the store, first under the
+// assumption that the input identifier is a model name, then using treating
+// the identifier as a full or partial model UUID.
+// If a model is successfully located its name and details are returned.
+func (c *ModelCommandBase) modelFromStore(controllerName, modelIdentifier string) (
+	string, *jujuclient.ModelDetails, error,
+) {
+	// Check if the model identifier is a name that identifies a stored model.
+	// This will be the most common case.
+	details, err := c.store.ModelByName(controllerName, modelIdentifier)
+	if err == nil {
+		return modelIdentifier, details, nil
+	}
+	if !errors.IsNotFound(err) {
+		return "", nil, errors.Trace(err)
+	}
+
+	// If the identifier is 6-8 characters or a valid UUID,
+	// attempt to match one of the stored model UUIDs.
+	l := len(modelIdentifier)
+	if (l > 5 && l < 9) || names.IsValidModel(modelIdentifier) {
+		models, err := c.store.AllModels(controllerName)
```

Comment:
Is this only the locally known models? will this be a problem connecting to a controller that has a model foo that someone else created that this client hasn't connected to yet?
(all of this code looks like it would suffer from that whether it was by name or by uuid, so maybe we don't care to fix that just yet).
Ideally, I think, we would resolve model names from the controller, but it probably is not easy to do so.

Reply:
`modelFromStore` shims the new logic in as a replacement for `store.ModelByName`, so at this site it only deals with the local store, but `modelDetails` does take care of a refresh-and-retry if the model is not found.